### ISSUE TITLE
Add `where` argument to `argmax`, `argmin`, `ptp`, `cumsum`, `cumprod`

### ISSUE THREAD

**Issue by carlosgmartin** [google/jax#20177]
The following functions receive a `where` argument, which limits the reduction to a given boolean mask:
- [max](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.max.html)
- [min](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.min.html)
- [sum](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.sum.html)
- [mean](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.mean.html)
- [all](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.all.html)
- [any](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.any.html)
- [prod](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.prod.html)
- [var](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.var.html)
- [std](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.std.html)

**Feature request:** Add a `where` argument to the following functions:
- [ ] [argmax](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.argmax.html)
- [ ] [argmin](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.argmin.html)
- [ ] [ptp](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ptp.html)
- [ ] [cumsum](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.cumsum.html)
- [ ] [cumprod](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.cumprod.html)

Related:
- https://github.com/numpy/numpy/issues/14371
- https://github.com/numpy/numpy/pull/21625

**Comment by jakevdp** [comment]
Hi - thanks for the request! We generally follow the NumPy API in `jax.numpy`, and as far as I can tell, numpy does not support a `where` argument for any of these functions.

**Comment by carlosgmartin** [comment]
@jakevdp Recalling what we did for [jax.numpy.fill_diagonal](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.fill_diagonal.html) with `inplace` (vs. [numpy.fill_diagonal](https://numpy.org/doc/stable/reference/generated/numpy.fill_diagonal.html)):
- https://github.com/google/jax/issues/2680#issuecomment-1763273560
- https://github.com/google/jax/pull/18180

I think that, since the API will be identical when this functionality is added to numpy (see the linked issue), implementing it in JAX ahead of time should not cause any issues.

Personally, it would be very helpful to have this functionality.

**Comment by jakevdp** [comment]
`fill_diagonal` is somewhat different: it's *impossible* to implement that in JAX without changing the API to avoid in-place modification.

Adding additional functionality to existing APIs is a qualitatively different discussion. There is some upside in flexibility, but it has some downsides too, namely:

1. It increases the user-facing API surface that we need to support for the rest of time.
2. It makes `jax.numpy` harder to test, because we must also implement the ground truth version to test against.
3. If NumPy or the Array API eventually adds the same keyword with different semantics, it makes for an awkward deprecation (see e.g. the `arr.device` method deprecation that we're doing currently).
4.  It increases, if only marginally, the cognitive burden of switching between `jax.numpy` and original `numpy`.
 
For those reason, I lean toward not adding these sorts of keywords until they are part of either the NumPy API or the Python Array API.

**Comment by carlosgmartin** [comment]
@jakevdp Thanks for your response! Is there a namespace for useful array functionality that has not yet been incorporated into NumPy? Perhaps [`jax.lax`](https://jax.readthedocs.io/en/latest/jax.lax.html) or some theoretical `jax.numpy_experimental`?

---

Somewhat off-topic, opinionated tangent:

It seems detrimental to JAX's evolution to tie itself so strictly to NumPy. 😔

This tying has also caused issues in the past, when NumPy's way of doing things is awkward.

Is there any existing theoretical discussion on the benefits, drawbacks, and future of JAX tying itself strictly to NumPy's API?

This is perhaps a provocative opinion, and purely speculative at this point, but I'd love to see JAX liberate itself from NumPy's straightjacket. 🙂 Especially if it eclipses NumPy in terms of userbase (which I'm hoping for).

For example, I find certain aspects of PyTorch's tensor API to be superior to NumPy's array API, from a usability perspective. (It's also worth noting that NumPy itself currently breaks with the Python Array API.)

**Comment by carlosgmartin** [comment]
@mattjj One example is how PyTorch makes it easier to chain array operations. Consider the following:

```python3
jnp.sqrt(jnp.square(abs(x.sum(2)).max(1)).sum(0))
```

Notice the "back-and-forth" or "spiral" pattern of the function calls. Compare it to

```python3
x.sum(2).abs().max(1).square().sum(0).sqrt()
```

which is much easier to write and read:
1. It follows the logical order of the operations.
2. It avoids using (and importing) `jnp.`, which reduces code clutter.

In my experience, this kind of pattern is common.

**Comment by jakevdp** [comment]
Thanks for elaboration. I've seen this kind of idea come up before in other contexts (e.g. https://github.com/numpy/numpy/issues/24081)

I personally don't feel strongly about this (it would be convenient in some cases, but that convenience comes with a cost) but I'm happy to hear what others might think.

**Comment by carlosgmartin** [comment]
Thanks for the link. It's indeed very relevant. It also brings up another advantage I forgot to include:

> With fluent API in common, users may more easily write library-agnostic code.

I saw this in practice here: https://github.com/cvxpy/cvxpy/issues/2237.

Anecdotally, since it was mentioned in that thread, I've personally also found pandas's method-chaining API a pleasure to work with.

Unfortunately, I think NumPy devs made the wrong call here. If I *had* to choose between standalone functions and methods, I'd choose the latter, due to the aforementioned advantages from the perspective of user experience.

**Comment by jakevdp** [comment]
I understand the choice the numpy devs have made. For what it’s worth, I wouldn’t suggest trying to re-open that discussion. They’ve made their opinions pretty clear, and it’s a very small team supporting a very large user community.

**Comment by carlosgmartin** [comment]
> I understand the choice the numpy devs have made.

But do you agree with it? Does the user community agree with it?

I won't try to steer NumPy's course; it looks like that ship has sailed. I'm more interested in JAX's future. (Again, speaking theoretically at this point.)

It might be wise to survey the community for these kinds of impactful, long-lasting design decisions. And to ensure surveyees have a good understanding of the issue, each alternative could be presented with a list of its pros and cons (perhaps even an argument map).

This approach, together with maintaining a dedicated space for open discussion of the alternatives for a reasonably extended period of time before committing, could be helpful to establish a confident, community-wide consensus.

Any thoughts welcome. 🙂

**Comment by jakevdp** [comment]
> But do you agree with it?

On the whole, yes I do.

I understand the reason that you are advocating for method-based access to all ufuncs and reductions: there are certain situations in which it could make for cleaner, more concise code. But I think the biggest danger to programming languages and DSLs in the long term is bloat in the API surface, and so it is prudent to be conservative when considering addition of new APIs or duplicate spellings of existing APIs.

NumPy is quite conservative in this way, and I think that's inextricably tied to why it's remained a successful project for a quarter century. On the JAX side, we have benefitted from using numpy (and more recently the Python Array API) as a standard, because it lets us take the prudent stance without having to debate the individual merits of every single proposed API extension. I believe that is good for JAX as a whole.

Is the JAX API perfect? No. Are we willing to change it? Yes, and we add features every day! But it would not be wise to approach changes like the ones proposed here without careful consideration of the costs as well as the benefits.

**Comment by carlosgmartin** [comment]
Just making a note of this here so I don't forget when https://github.com/numpy/numpy/issues/26336 is completed:

- Once we add `where` to [`argmax`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.argmax.html), we can add `where` to [`random.categorical`](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.categorical.html) (which uses `argmax` internally), removing the need to manually mask logits with `-jnp.inf`.

**Comment by carlosgmartin** [comment]
@jakevdp While this is being worked on from the NumPy side, in the meantime, would it be acceptable to add an optional mask argument to [`lax.argmax`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.argmax.html), which [`numpy.argmax`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.argmax.html) uses internally? That way, as soon as NumPy adds the `where` argument, JAX can quickly do the same.

**Comment by jakevdp** [comment]
`lax` functions by design are more-or-less a direct wrapper for the underlying XLA operation; I don't think it's appropriate to add masking support there when the XLA op doesn't implement it.

**Comment by carlosgmartin** [comment]
I didn't find an XLA op for argmax. Is the argmax reduction defined in JAX itself?

I found the following chain of definitions: [argmax](https://github.com/google/jax/blob/main/jax/_src/lax/lax.py#L1017) > [argmax_p](https://github.com/google/jax/blob/main/jax/_src/lax/lax.py#L4193) > [_compute_argminmax](https://github.com/google/jax/blob/main/jax/_src/lax/lax.py#L4176C23) > [_ArgMinMaxReducer](https://github.com/google/jax/blob/main/jax/_src/lax/lax.py#L4152).

**Comment by jakevdp** [comment]
`lax.argmax` is essentially a convenience wrapper around `lax.reduce`
